### ISSUE TITLE
Google Cloud BigQuery: API may change

### ISSUE THREAD

**Issue by ennru** [akka/alpakka#2353]
Alpakka Google Cloud BigQuery was added in Alpakka 2.0.2 in July 2020 and is marked as "API may change". Please try it out and suggest improvements.

**Comment by armanbilge** [comment]
Thanks for this addition!

Is there any interest in more user-controlled unmarshalling? I use Circe in my project, so right now in the `parserFn` function I am first converting from Spray's `JsValue` to Circe's `Json` before parsing into my desired type, which is less than ideal.

I was expecting an interface more like Akka HTTP, which is very flexible with regard to marshalling/unmarshalling. Is there a reason this is tricky to implement? I'd be happy to take a stab at this in a PR.

**Comment by ennru** [comment]
Other than trying to keep it simple there is no reason to require Spray JSON. If you can suggest a good way to allow for other JSON parsers please do.

**Comment by armanbilge** [comment]
Thanks for the response! My suggestion is to reuse Akka's existing Unmarshalling infrastructure that is already used by Akka HTTP. There are many libraries that implement Unmarshallers via a number of JSON parsers, including official Akka libraries that use Spray. In fact, the current implementation is [already using the Unmarshalling API](https://github.com/akka/alpakka/blob/master/google-cloud-bigquery/src/main/scala/akka/stream/alpakka/googlecloud/bigquery/impl/parser/Parser.scala#L72), just in a hard-coded manner.

More specifically, instead of asking the user to provide a `parserFn: JsObject => Try[T]` when creating a `GoogleBigQuerySource`, we ask for an (implicit) `FromEntityUnmarshaller[T]`.

Does that make sense? Thanks again for your consideration.

**Comment by armanbilge** [comment]
Aha, now that I looked more into this I understand what makes it tricky. The `GoogleBigQuerySource` implementation itself needs to retrieve the `pageToken` and `jobId` from the JSON. So to support custom Unmarshalling means that either the user must also parse and return the `pageToken` and `jobId` or the `GoogleBigQuerySource` implementation needs to parse the JSON itself in addition to the user-provided Unmarshalling. So perhaps there isn't a good way to abstract out the JSON implementation after all.

**Comment by mrooding** [comment]
I was just asking @ennru a similar question about having the ability to switch JSON libraries for the Elasticsearch extension. It would be ideal if we could come up with 1 solution that's applicable to both the Google Cloud and Elasticsearch alpakka extensions.

A possible solution is to use type classes. This'll allow us to create implementations for different JSON libraries. The user would then only need to bring the right one into scope and provide the library-dependent type information for the types that will be read/written (e.g. `Decoder` and `Encoder` for circe).

To limit the impact on alpakka users, we'd need to see if we can achieve this setup that supports a default implementation (probably set to spray-json?).

I'm willing to pitch in to set this up and I'm curious to hear what the maintainers of alpakka think about the proposal, or whether they see a better alternative.

**Comment by armanbilge** [comment]
Glad to know I'm not the only one needing this! I agree that type classes are the way to go, my vote is to use the existing `Marshaller` and `Unmarshaller` type classes already in Akka HTTP if possible. Then users can use the appropriate module from [akka-http-json](https://github.com/hseeberger/akka-http-json) to plug-and-play with the JSON library of their choice. Akka provides official support for (un)marshalling with Spray via the `akka-http-spray-json` module. Thoughts?

**Comment by mrooding** [comment]
If we can get that done that would be amazing since we'd be able to utilise existing extensions like https://github.com/hseeberger/akka-http-json. I'm not too familiar with the Google BigQuery API but you mentioned in your previous comment that it may not be a good idea to abstract out the JSON abstraction. Can you elaborate on that?

When I've got a bit more time I'll play around with the Elasticsearch code to see if I can get Marshaller and Unmarshaller support to work.

**Comment by armanbilge** [comment]
Exactly! The marshalling/unmarshalling API is already well-supported by the Akka ecosystem.

Basically, the BigQuery API provides a JSON response that [looks like this](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/getQueryResults):
```
{
  "jobReference": {
      "projectId": string,
      "jobId": string,
      ...
  },
  "pageToken": string,
  "rows": [ ... ],
  ...
}
```

As a user of the Alpakka API, most likely\* I am interested in building a  `Source[T]` that provides unmarshalled instances of the object of type `T` encoded in the `rows` key. However, to implement this source the Alpakka API needs the `jobReference` and `pageToken` to request the next page of results from BigQuery. Ideally, this process should be completely hidden from the user.

JSON libraries generally implement parsing as:

```
string -> library-specific JSON type -> user's desired type
```

So to support interchangeability of the JSON library, Alpakka would have to rely on the user to parse and provide `jobReference` and `pageToken`. IMO, I think this level of interdependence makes for a brittle API.

Currently, Alpakka parses the JSON response itself into Spray's `JsObject` type and retrieves the `jobReference` and `pageToken` keys before passing that `JsObject` on to the user. To support user-provided unmarshalling, Alpakka could discard the `JsObject` and instead pass the raw JSON response (i.e., a string) to the user to parse again with their desired library. Obviously, this is expensive (double-parsing of the same JSON) and wasteful in the case that the user is using Spray anyway!

Let me know if that makes more sense. Not sure if this is relevant to the Elasticsearch API at all?

\* For the sake of brevity, I omitted a number of keys from the BigQuery JSON response, such as `schema`, `totalBytesProcessed`, etc. A more sophisticated user of the BigQuery and Alpakka APIs might in fact need access to the entire JSON response and not just the contents of `rows`. So this would have to be supported as well.

**Comment by seglo** [comment]
With the #2548 rewrite I think we can close this ticket.
